### ISSUE TITLE
Outfitter Screen Is Convoluted and Inconsistent

### ISSUE THREAD

**Issue by SimulatedKnave** [endless-sky/endless-sky#7888]
### Problem Description

I'm always frustrated when using the outfitter screen. The same hotkeys do different things depending on which checkbox or ship is selected, and very basic things one might expect to be able to do (move a module from cargo to storage and then back, for example) are surprisingly involved. Buying to cargo is 100% impossible while...looking at your cargo. I didn't realize you could do it until after I'd beaten the FW campaign. Which was about three or four starts in. 

I literally can't remember all the various combinations of checkboxes and buttons to do particular things. That's not a good sign. 

There is, as far as I can tell, no reason to not simply have the outfitter screen work as "push button to do X with what you have selected." 

B can buy. 
S can sell (prioritizing cargo in the cargo screen and storage in the storage screen, as I believe it currently does). 
I can install. 
U can send an item to storage (uninstalling it if necessary)
C can send an item to cargo (uninstalling it if necessary)
Notably, ALL these buttons DO do those things at some times, just not consistently. C, for example, sometimes sends stuff to cargo and sometimes sells. U is uninstall, but is also how you move cargo to storage. 
Alternately, have T send stuff to storage, and U always uninstalls. 

Have cargo be sold if cargo is selected, storage if storage is selected, and if there's nothing in cargo or storage then the installed one is sold. If there are actually buttons to move stuff from cargo to storage easily, that's MORE than flexible enough to let a player do whatever the hell they need to (especially if T sends to storage while U always uninstalls). 

Also, I think almost all of how it DOES work is undocumented, which doesn't exactly make it easier. Though God knows I wouldn't want to try to document it. 

### Related Issue Links

#7253 

### Desired Solution

The same buttons should always do the same things while in the outfitter screen. At minimum, that should be the case. It almost doesn't even matter what they do. Just...the same thing. 

### Alternative Approaches

...There is no valid alternate approach I can think of. I have no idea how this situation came to be. It is possible there is a good reason. I cannot conceive of it.

**Comment by bene-dictator** [comment]
You can get shortcut tooltips to stay up all the time now in the continuous version, so you can see what some of these are. But I agree.

**Comment by yafflehk** [comment]
Should there be separate screens for storage and cargo? The current system feels a lot like it was cobbled together by just adding more and more functions to the outfits screen.

**Comment by SimulatedKnave** [comment]
I think having it be shared is workable enough. Adding another checkbox for 'installed' might be a good way to ensure the player has complete control over whatever they're trying to screw around with, though. There really is a lot of advantage to being able to look at all the stuff at once.

**Comment by ThrawnCA** [comment]
What if there were modifier keys that changed Buy and Sell into different actions (including visible change)? Eg holding 'c' could change them from simple "buy+install" and "uninstall+sell" into "Buy into cargo" and "Sell from cargo". Holding 'c'+'u' could change them to "Move from storage to cargo" and "Move from cargo to storage".

**Comment by SimulatedKnave** [comment]
If that is in addition to my suggestion, sounds interesting (especially the move between two points bit). 

If that is instead of...I'm pretty sure mine is more straightforward. Indeed, if you add a "view installed outfits" view I think it covers every possible thing you might want to do.

**Comment by ThrawnCA** [comment]
The advantage of a modifier key setup is that you don't need to have separate views, and even if you did still have them, modifier key behaviour would not depend on them; it would let you consistently perform any action at any time.

It would also simplify complex cases such as sending an item to storage when you have one installed and also have some in cargo. Which one should go first? Well, if you press 'i+u' (installed+storage) and sell, then the currently installed outfit gets uninstalled and goes to storage, whereas if you press 'c+u' (cargo+storage), then one from cargo goes to storage. And if you just press 'u', then a new outfit is purchased into storage. There's no ambiguity.

**Comment by SimulatedKnave** [comment]
Except separate views are actually useful, by allowing the player to not look at stuff they don't want to actually look at. Sometimes I just want to see what's in my cargo and nothing else. 

Also, I am 99% certain that "add a key, have it do X all the time" is easier than "have a bunch of keys do different things at different times, and other things when in combination."

**Comment by Zireael07** [comment]
Both of your opinions have definite pros. A modifier key setup might be easier to extend code-wise, BUT is more easily missable by the players.

**Comment by ThrawnCA** [comment]
> Except separate views are actually useful, by allowing the player to not look at stuff they don't want to actually look at. Sometimes I just want to see what's in my cargo and nothing else.

As mentioned, views could still exist, and the default behaviour could even continue to change based on view. But you would also be able to perform operations using a consistent key combination, regardless of which view you're on.

> Also, I am 99% certain that "add a key, have it do X all the time" is easier EDIT: to code/EDIT than "have a bunch of keys do different things at different times, and other things when in combination."

Depends what you mean by easier to code. Having a simple yet extensible concept - use input to choose two points from a list, then perform an operation between those two points - is certainly more _elegant_ to code than having a list of "if pressed A, do X; else if pressed B, do Y; else if pressed C, do Z." Easier to read and to reuse.

It's conceptually similar to what already happens with the numeric modifier keys, and it's comparable to the keybinding approaches used by vi and emacs, which are definitely programmer-oriented.

I'm definitely not arguing against the idea that keys should have consistent meanings, I'm just suggesting that there's an opportunity to revamp what the keys do so it properly takes advantage of that consistency to make things more efficient.

**Comment by ThrawnCA** [comment]
> Indeed, if you add a "view installed outfits" view I think it covers every possible thing you might want to do.

Hmm. By my count, there are currently 12 possible operations. If there are two operations available on each view, and a fourth view was added, that's 8 operations, maybe ten if there's extra components specifically added for moving things between cargo and storage. So there must be something missing, something that is more complex to do.

Let's see...

| Operation | View-based solution | Modifier key solution | Notes |
| ------------- | --------------------------- | --------------------------- | -------- |
| Buy new outfit and install | Buy with installed outfits visible but not cargo? | i + Buy | Not sure if this can be done reliably via views when both installed and cargo versions of the outfit exist |
| Uninstall outfit and sell | Sell with all outfits visible? | i + Sell | Not sure if this can be done reliably via views when both installed and cargo versions of the outfit exist |
| Buy new outfit into cargo | Buy on cargo view (cargo visible but installed outfits not visible) | c + Buy | |
| Sell outfit from cargo | Sell on cargo view | c + Sell | |
| Buy new outfit into storage | Buy on storage view | u + Buy | Storage view: storage visible but installed and cargo not visible |
| Sell outfit from storage | Sell on storage view | u + Sell | |
| Install outfit from cargo | Buy with installed and cargo visible | i + c + Buy | |
| Uninstall outfit to cargo | ? | i + c + Sell | |
| Install outfit from storage | Buy with installed outfits and storage visible | i + u + Buy | |
| Uninstall outfit to storage | 'u' with installed outfits visible | i + u + Sell | |
| Move from storage to cargo | Proposed new function | c + u + Buy | |
| Move from cargo to storage | Proposed new function | c + u + Sell | |

Anything I've missed or mistaken?

**Comment by SimulatedKnave** [comment]
"Buy new outfit and install" can be duplicated effectively by "buy new outfit to storage" + "install." Or the other way around, for that matter (i.e. how the game does it if you don't untick the cargo box). 

Also, move from storage to cargo and cargo to storage are existing functions. You just have to untick counterintuitive boxes to get them (you can move stuff from storage to cargo by unticking the "view cargo" box, then pressing C on whatever you have selected. And cargo to storage is just pressing U). 

The way I'd see it is as a hierarchy of checkboxes. You have: 
For sale
Storage
Cargo
Installed

Selling and installing pull from the highest checkbox currently ticked (or the next one if there isn't anything that fits the highest one). Then you have the buttons to transfer what you have selected around between installed/cargo/storage. 

This is sort-of how the game does it now, it just is a lot trickier to predict, and the default is installing everything, which is a pain.

**Comment by ThrawnCA** [comment]
> "Buy new outfit and install" can be duplicated effectively by "buy new outfit to storage" + "install." Or the other way around, for that matter (i.e. how the game does it if you don't untick the cargo box).

Are you saying that it will take two steps to perform one action? Not ideal.

> Also, move from storage to cargo and cargo to storage are existing functions. You just have to untick counterintuitive boxes to get them (you can move stuff from storage to cargo by unticking the "view cargo" box, then pressing C on whatever you have selected. And cargo to storage is just pressing U).

Good to know, though as you say, it is counterintuitive to make a view vanish in order to send items there. Also, it seems probable that in such a scenario, the Buy and Sell buttons are not opposites of each other? If at all possible, they should be.

> The way I'd see it is as a hierarchy of checkboxes. You have: For sale Storage Cargo Installed
> 
> Selling and installing pull from the highest checkbox currently ticked (or the next one if there isn't anything that fits the highest one).

That sounds somewhat like my suggestion, but using views instead of keys. It's the equivalent of holding a single modifier key.

> Then you have the buttons to transfer what you have selected around between installed/cargo/storage.

This, however, is not very clearly defined. What buttons would you envisage seeing if, for example, you have checked Storage and Cargo but not Installed?

**Comment by SimulatedKnave** [comment]
> Are you saying that it will take two steps to perform one action? Not ideal.

...Hit B, then hit I. Less involved than a modifier key, for that matter. 

> Good to know, though as you say, it is counterintuitive to make a view vanish in order to send items there. Also, it seems probable that in such a scenario, the Buy and Sell buttons are not opposites of each other? If at all possible, they should be.

I don't follow what you mean here by the buy and sell buttons being opposite of each other. 

> That sounds somewhat like my suggestion, but using views instead of keys. It's the equivalent of holding a single modifier key.

Yes, except holding a modifier key requires me to remember the key and hold the key down. While holding a key down isn't the end of the world, it can get tiring if you're doing it for a while (or depending on key location). 

> This, however, is not very clearly defined. What buttons would you envisage seeing if, for example, you have checked Storage and Cargo but not Installed?

...The same buttons as always. T for storage, C for cargo (I suppose you could still have I install). The views allow you control over what you pull from more than over where it ends up (which, again, is how it works currently. It's just ungodly confusing right now).

**Comment by ThrawnCA** [comment]
> ...Hit B, then hit I. Less involved than a modifier key, for that matter.

Not really? Same number of keystrokes, but requiring you to make sure you're in the right view first.

> I don't follow what you mean here by the buy and sell buttons being opposite of each other.

I mean that if you click Buy and then click Sell, ideally you should be in exactly the same position as when you started.

If you have the Storage view open, you can click Cargo (the replacement for Buy), but you can't just click the other button to reverse the operation. Indeed, if it was the last one in storage, then it's no longer even visible.

> Yes, except holding a modifier key requires me to remember the key and hold the key down. While holding a key down isn't the end of the world, it can get tiring if you're doing it for a while (or depending on key location).

It's already used for numeric modifiers, with up to three modifier keys at once.

> > This, however, is not very clearly defined. What buttons would you envisage seeing if, for example, you have checked Storage and Cargo but not Installed?
> 
> ...The same buttons as always. T for storage, C for cargo (I suppose you could still have I install).

Not keys; I meant, Which buttons should appear on screen?

That highlights another advantage of a modifier key approach: any keypress that results in an actual change, corresponds to a visible button that you could click. Whereas having the ability to press 'T' to send to storage or 'C' to send to cargo does not.


Hmm. How feasible do you think it would be to combine the two approaches by altering the current checkbox behaviour?

Currently, the default state is that all checkboxes are selected, and you deselect them as needed. That's okay for just choosing visibility, but as a way of selecting where to move outfits around, it's horribly counterintuitive. What if it were reversed, so that the default state is "No checkboxes selected, everything is visible" and you only select the checkboxes when you want to limit the view to that category? So, checking the Cargo box would shrink your view from "everything" to "just outfits in cargo".

The checked views could then modify the Buy and Sell buttons in the same way as my proposed modifier keys. Select just one, you're moving outfits between that one and the outfitter. Select two, you're moving between them. Select more than two, one is ignored.

**Comment by SimulatedKnave** [comment]
> Not really? Same number of keystrokes, but requiring you to make sure you're in the right view first.
> 

First, the existing system already requires that. Second, if the views all default to on, that's perfectly workable. 

> If you have the Storage view open, you can click Cargo (the replacement for Buy), but you can't just click the other button to reverse the operation. Indeed, if it was the last one in storage, then it's no longer even visible.
> 

Cargo isn't a replacement for Buy. 

Pushing C would move stuff to Cargo. Pushing T would move stuff in Cargo to storage. So pushing the two keys as you describe would in fact put you where you were before. 

> It's already used for numeric modifiers, with up to three modifier keys at once.
> 

And it's annoying there too. 

> > > This, however, is not very clearly defined. What buttons would you envisage seeing if, for example, you have checked Storage and Cargo but not Installed?

I am using the term button to mean keys. While I admit that that is perhaps a fault on my part, the bit where I specifically name letters in the initial post would have, I thought, made it clear. As well as me talking about what they currently do. 

I am not suggesting adding a bunch of clickable buttons to the interface. 

> The checked views could then modify the Buy and Sell buttons in the same way as my proposed modifier keys. Select just one, you're moving outfits between that one and the outfitter. Select two, you're moving between them. Select more than two, one is ignored.

That sounds more complicated, frankly. And significantly less intuitive. "This is checked, you are looking at the stuff that is checked" is logical. "Everything is unchecked, but check what you want to look at specifically" is not. 

The behaviour you describe above could be duplicated simply by having checked boxes work the sensible way (more-or-less as they currently do view-wise, just with interaction also following the view), then unchecking boxes.

**Comment by SimulatedKnave** [comment]
I have taken a brief look at the source code. Assuming I am understanding it correctly (not a given) this is simultaneously fairly straightforward and hugely involved. It is fairly straightforward because the outfitter code seems fairly clearly written, and so it would be relatively straightforward to just add a third 'installed' category for items to be in, then add moving stuff around between them. 

Is there currently a simple condition that can be used for things in cargo but not installed? I know the code CAN make the distinction, I'm just not sure what the term it uses for "cargo, not including installed items" is. Or perhaps everyone's just using cargo and specifically excepting installed stuff in those scenarios. Or perhaps I've misunderstood. 

Assuming it DOES allow for that, if someone tells me the right terms to use for "in the cargo hold" vs "installed or in the cargo hold" (instead of just "cargo," which seems to be catching both) I'd even be prepared to draft something. Just no one's allowed to complain at how godawfully buggy it'll probably be.

**Comment by ThrawnCA** [comment]
Honestly, I think that in order to properly continue with this conversation, I'd have to sit down for an hour or so with the existing outfitter screen to try out all the permutations and lock down exactly what can do what. Because, as per the issue title, its current state is complex and unintuitive and often hard to predict.

> "This is checked, you are looking at the stuff that is checked" is logical. "Everything is unchecked, but check what you want to look at specifically" is not.

It's logical enough if you think of the checkboxes as search filters. With no filters, everything is shown.

**Comment by SimulatedKnave** [comment]
Fair. That said, I think the checkboxes work better in their current setup (looking at the code, I'm not sure how easy it would be to change them, either).

**Comment by ThrawnCA** [comment]
So, I discovered this morning, while fiddling with things, that if you have an outfit both installed and in cargo, then unticking Storage (leaving the others ticked) will make 'u' move an item from cargo to storage, instead of uninstalling. Which is handy to have, and nice to know, but has zero logic to it that I can follow.

I'm pretty sure that it would be more logical and intuitive to redesign it so that when 'cargo' and 'storage' are the only ones checked, the Buy button becomes "Move from storage to cargo" and the Sell button becomes "Move from cargo to storage".

**Comment by johndh** [comment]
Just to offer my two cents from the mobile device crowd, a system relying on modifier keys doesn't translate well to a touchscreen, while a tabbed view or check boxes both seem fine to me.

Related: https://github.com/thewierdnut/endless-mobile/issues/24

**Comment by TheFurnace** [comment]
Modifiers would be an issue for me since there's already a quantity modifier. I like the idea suggested that there be a set of simple steps: select item -> select view -> select action (modified by quantity).

**Comment by TheRemyLeBeau** [comment]
My apologies for barging in here talking like I know what I'm talking about, but I'm fairly new at this game. Perhaps my perspective would provide some insight that veterans don't take into account because they're used to how the game works and has worked for years. If not, feel free to call me out for being a moron ;)

**Filters**

What I'm seeing in this discussion is how hotkeys are inconsistent, because they're dependant on which filters you have active. A key does different things depending on what filters are active. That's very confusing.

Now, once again my apologies if I sound stupid, but shouldn't filters be a ... _visual limitation_ rather than _changing_ what the same key/button does? 

**Install Function**

Sure, the idea of replacing the 'Buy' button with 'Install' sounds pretty logical initially, because the thing you want to do at an outfitter is to buy something to install it on a ship. That's most likely what 90% (arbitrary number to show my ignorance) of the people do at an outfitter, so _combining_ that by replacing the 'Buy' button to 'Install' seems smart. It's catering to what people want and removing additional frustrating steps.

But this becomes a problem when you do _not_ want to install it.

**Inventory management**

With the option to buy things without installing it on the ship, you technically fall into the 'inventory management' rabbit hole. That allows you to put it in 'cargo' (where it occupies space on the ships), 'storage' (from what I read it's in the outfitter storage, not on display to sell, which wasn't clear to me) or 'in stock' (allowing it to be sold by the outfitter).

Based on the discussion above there's a debate on how to make it more consistent where stuff goes, which is perfectly valid to me. But it seems like we're trying to find solutions confined to the already existing setup, with minor changes left and right to patch things up to make it feel less awkward. Wouldn't it be better to just ... look at what we actually need?

**Proposal**

We already have filters. We also have a search button. Why not simply introduce a checkbox-system where you can pick 'Move to cargo' and 'Move to storage' on the right side of the search button? This could solve a couple of things.

1. If you check this box, it will be put into either cargo or storage. That way the 'Buy' button can stay a 'Buy' button rather than an 'Install' button, unless you untick both boxes (default setting), which means it's effectively an 'Install' operation. This means you can buy items intuitively through the GUI without having to resort to hotkeys or deselecting your ship. (Fun fact: the game doesn't tell you _deselecting your ship_ is a requirement to buy stuff without installing through the GUI, nor _how_ you manage this, so this proposal would solve multiple issues on that front).
2.  The entire discussion where stuff would go based on hotkeys would be much easier, because the hotkeys only have to verify these checkboxes to decide which function to execute. This would make it ultimately more consistent _and_ easier to understand for the player. 
3. It could _possibly_ also make the Outfitter code shorter or clearer because you'll only need a dependency on these two checkboxes to decide which function to execute.
4. There's no need to discuss 'priorities' that take place where something goes, because the player decides where it goes based on a single variable. That could potentially also clear up code and optimalize things.
5. Ticking a checkbox means you can do multiple purchases in the same manner, without having to fiddle around.

**Issues**

Now, I already see a couple of issues resulting from this. 

1. Checking _both_ checkboxes ('in cargo' and 'in storage') would result in odd behavior. So you'll have to make them mutually exclusive. Usually that means you resort to radiobuttons, but I'm not sure how viable a radiobutton would be in this setup which would require a third ('Install') option as well.
2. I'm not familiar enough with different OSes that people use. I'm seeing people playing this game on mobile devices, which could be a dealbreaker for additional checkboxes or radiobuttons on an already limited screen compared to a PC.

**Bottom line**

I hope you can appreciate my perspective and my proposal. Even if it's not a viable solution, perhaps it helps in gaining new perspective to see new options.

**Comment by ThrawnCA** [comment]
Check boxes for cargo and storage would be a nice solution for _part_ of the problem. But I think they would leave some other parts of the problem untouched - while taking up valuable screen real estate.

Of the twelve operations I identified, check boxes would handle something like half very neatly. Buy+install, buy into cargo, buy into storage, sell from storage, sell from cargo, uninstall+sell. But they wouldn't handle the other half, things like "uninstall into cargo" or "move from cargo to storage". Unless I'm misunderstanding?

**Comment by TheRemyLeBeau** [comment]
> Of the twelve operations I identified, check boxes would handle something like half very neatly. Buy+install, buy into cargo, buy into storage, sell from storage, sell from cargo, uninstall+sell. But they wouldn't handle the other half, things like "uninstall into cargo" or "move from cargo to storage". Unless I'm misunderstanding?

The way I envision it, 'Move to cargo' could be applied to this as well. If you check this box, it would:

1. Buy an item and move it to cargo rather than installing it.
2. Move an item from storage and move it to cargo.
3. Uninstall it and move it to cargo.
(Phrasing may be a bit awkward, but I tried to be consistent with the term 'move to cargo' I used before, while 'put into cargo' probably sounds better in hindsight.)

After all, the checkbox says: "Move to cargo", so it should be applied to everything you select and use an action for. Whether it comes from the store, your ship or storage shouldn't matter since all of these actions require the outfit to be moved to cargo. 

Considering there's a second checkbox 'Move to storage', that should cover the other half of your options like 'Move from cargo to storage' and 'Uninstall to storage'.

**Tricky**

Where it gets tricky is when you select _new_ items from the store and already _existing_ items in storage or cargo. Since they're different sources, this might require a bit of intelligence from the code. I don't know how the current code copes with this, but I assume there's already something in place to cover this. Just figured I'd point it out in case there isn't (in which case it probably counts as a bug).

**Buttons**

This means there's no need for additional buttons next to the 'Buy', 'Sell' and 'Leave', so that part of the screen does not require new space. One does have to check whether the naming is still consistent when applying this. 'Buy + install' changes the 'Buy' button to 'Install'. 'Uninstall' is also a name change for the 'Buy' button. I don't think 'Buy' needs to change when you buy something and move it to cargo, because it's still a buy action.

How do we want to approach this when moving stuff from cargo to storage or the other way around? Would that same button be intuitive, or would we rather make use of the 'Sell' button instead? I don't have enough experience with these actions to make an educated guess at this point. 

I assume the button label could also be rephrased to 'Cargo' or 'Storage' when you intend to move stuff around to clarify what the button does in this instance. You could even consider the universal 'Move' label if that's more intuitive.
(That could also be a nice warning when you _unintentionally_ tick the box and see the button name has changed.)

*Hotkeys*

I'm not quite as familiar with the hotkeys for the Outfitter screen, so applying hotkeys for this wasn't part of my equasion. Considering you've already made a list for all possible actions with hotkeys, do they still work in my proposed solution or would you reconsider the hotkey combinations for that? But that's probably a discussion you should have with @SimulatedKnave rather than with me ;)

**New/Casual Players**

As I said before, new and casual players (like me a dozen hours ago) usually want to sell whatever they don't need since they either don't have the fleet (or don't envision to having to save outfits to install on a future fleet), so the default should just be selling items when uninstalled and installing when buying. That system is already in place, so that's a plus.

The filters have an obvious naming standard already, so that shouldn't interfere with this.

**Comment by ThrawnCA** [comment]
Hang on, are you still depending on the current view-switching behaviour, with the checkboxes for cargo etc, and your proposed new checkboxes are just helping to define the target?

Firstly, I think it's agreed that the view-switching behaviour in its current state is remarkably confusing for new players. Secondly, having two sets of checkboxes with similar options will assuredly result in people mixing them up.

**Comment by TheRemyLeBeau** [comment]
>Hang on, are you still depending on the current view-switching behaviour, with the checkboxes for cargo etc, and your proposed new checkboxes are just helping to define the target?

At first glance I would've said 'yes', but you seem to imply that I want to keep the functionality in the original checkboxes and use the second checkbox in conjunction to specify a destination. That's not entirely correct. 

So let's try to phrase my idea in a very direct way to clear up some confusion:

1. The existing checkboxes are to be used to _display_ what's on screen. These checkboxes do _not_ influence the functionality of buying or selling. They're merely a _filter view_ to decide what you can see on screen. 
(From your phrasing: yes, 'view-switching' is correct'; 'behavior' is not correct, because a 'view' should be 'dumb' in this scenario.)
2.  The new checkboxes in my proposal will take over the actual functionality/behavior. Whether it's bought + installed on the ship, bought to cargo, bought to storage, moved from cargo to storage - all of the 12 options you've listed - these new checkboxes will determine how that works.

**Both sets of checkboxes**

So yes, I do say that we need both sets of checkboxes, because they serve different purposes. One is for display, the other for (destination) functionality. 

If you say having both will make it confusing, you can always opt to remove the original ones. That means you will _always_ have to display the outfits in store, cargo and storage, because you don't have any filters. 

Considering those are already existing functions, I'm unsure if that's the right approach. I suspect that people have requested to be able to filter outfits at some point, which is when these checkboxes got implemented initially. 

**Labels**

If your concern is that people will mix them up, then we should talk about what's the proper label instead, so it becomes clear to the player what these checkboxes do. Currently we have:

1. Show outfits for sale
2. Show outfits in cargo
4. Show outfits in storage

The text on these options seem pretty clear that they're purely for displaying stuff.

My initial phrasing for the new checkboxes were (I've added the word 'outfit' now to make it consistent with the original ones):

1. Move (outfit) to cargo
2. Move (outfit) to storage

Personally I think those phrases should be clear enough to understand the difference between the difference in functionality. I could be mistaken, though. I'm open for suggestions to rephrase these, if that's the main concern of using both sets of checkboxes.

In particular the word 'move' should indicate that it's not going to be installed on your ship. The words 'cargo' and 'storage' should reinforce that. As I've said before, 'put into' instead of 'move' could be used as well, if that's more intuitive.

**Comment by ThrawnCA** [comment]
I'm glad to see this is being improved, though I still think my suggested approach would have achieved a better result without adding a new 'R' key.

- Hold 'c': Buy and Sell buttons become "Buy into cargo" and "Sell from cargo"
- Hold 'u': Buttons become "Buy into storage" and "Sell from storage"
- Hold 'i': Buttons become "Buy+install" and "Uninstall+sell"
- Hold 'c'+'u': "Move from storage to cargo" and "Move from cargo to storage"
- Hold 'c'+'i': "Install from cargo" and "Uninstall to cargo"
- Hold 'u'+'i': "Install from storage" and "Uninstall to storage"

It covers every possible transfer, without relying on any context-based priority ordering, and conceptually it's similar to the existing setup of stacking x5/x20/x500 modifiers.

**Comment by xobes** [comment]
> ... without adding a new 'R' key.

The `R` key already existed in the code for selling ships and retaining the outfits in storage. I simply exposed that behavior in the Shipyard and used it as well in the Outfitter.

> Hold 'c': Buy and Sell buttons become "Buy into cargo" and "Sell from cargo"

How does one "Hold" the `c` key and... What's your keyboard shortcut for "Buy" then?

> Hold 'c'+'i': "Install from cargo" and "Uninstall to cargo"

Are you saying that how does one declare what direction things would go in the case where things could go either way? And would I have to hold `c` + `i` + `alt` + `ctrl` + `shift` .... and then another key, presumably to indicate that I'm ready to actually perform the action?

Some keyboards would balk at having 5 modifiers and letter keys are far less likely to be wired in a way to even allow them to be modifiers.

Finally: The need to support every single combination is not a true need and compromises were made to allow for single-button actions to continue to perform typical actions. My goals are stated clearly at the top of #10743 along with my analysis and **requirements**.

The two actions that you have identified as not being supported by #10743 are:
**Buying to storage** (today, where shipyards have unlimited stock) does not afford the player *any* benefit whatsoever, as retaining items in stock rather than in the shipyard is really just a matter of wasting credits.

With respect to **uninstall into cargo** does seem useful and is not supported as a direct pathway in the current implementation as "Uninstall" is akin to "Unload" in the current design and differs from sell only in that no credits are exchanged. However, in the case that you want to perform this action, there are a *couple* ways that this can be readily performed. The nominal mechanism for "Uninstall" can be thought of as "Unload" -- Remove from ship, no credits exchanged -- which will place outfits into storage. _I believe this was the previous behavior for both cargo and storage_. To achieve this outcome:
 - Option 1: Sell and re-buy (items sold ought to be re-purchased at the same price they were sold for) into cargo: First `S` then `C`.
 - Option 2: Uninstall and then move to Cargo from Storage: First `U` then `C`.

Note: the new code does provide a mechanism to support transfers from every *source* to every *sink* but what's missing is a way to override the GUI via a plugin and support for user-specified keys as well.  These things _may_ eventually come, so don't lose hope just yet. Consider this step one. Play it a bit, if you truly hate it, hope a new issue and we can further the discussion there.

**Comment by xobes** [comment]
Response to https://github.com/endless-sky/endless-sky/pull/10743#issuecomment-3694303620:

(I'm trying to consolidate to one thread since we're jumping back and forth here):

> Partially true. @xobes already identified a case where they are different, where you want to store more than your cargo space can fit. And with combinations of existing keys, the 'R' key would not be needed.

See above. Repeat: At this time there is no useful case to "buy" into "storage" as the price now will be the price later and there is infinite supply.

> But yes, it still requires multiple key presses. Feels different to me because it's ultimately one action, nothing changes until the key combination is complete.

See below, I'll refer to this in a minute.

> Also, being able to specify exactly what move you want to make means that no priority table is needed. If there are outfits both in cargo and storage, you don't have to try to remember which one will be used first to install. 

Fair.

> You just select the one you want: c+i to install from/uninstall to cargo, u+i for storage. If you want to sell from storage without touching cargo, you don't need to remember that cargo will be sold first, and installed outfits will be sold last; you just hold 'u' and press Sell.

See below, "press Sell" is the sticking point here. Along with up to 5 modifiers, see [earlier reply](https://github.com/endless-sky/endless-sky/issues/7888#issuecomment-3694924703) re: keyboard construction :-)

> It's easy for key combinations to give you visual feedback, changing the Buy and Sell button text, before you commit to an action. And it opens new opportunities for stacking with the numeric modifiers; you could, for example, use Alt+u+s as a shorthand for "sell all from storage without touching cargo" regardless of how many are in storage.

OK, this is where I repeat myself, hopefully concisely.

My requirements going into this were to change things from baseline as little as possible retaining the **single keypress to perform an action** that was previously implemented.  Your proposal expects a *button to be pressed* rather than supplying hot keys for all actions. (I, for one, don't use my mouse very much in the outfitter and rely on keyboard navigation wherever possible.)
 - Strike one: No keyboard hotkeys to perform actions.

Secondly you suggest "visual feedback, changing the Buy and Sell button text" which *masks the available functions* and was specifically one of the requirements that I had from the start. Available functions should be readily exposed to show users what actions are available rather than to change the function of buttons. I could argue that one button would be enough ("Do Action!") and every intent could be specified by modifiers. There is a line.
 - Strike two: No visual representation of what actions can be performed (hidden functionality).

> I would also argue that an approach where every action has a clear inverse, with the Buy and Sell buttons changing to different modes but always remaining exact opposites of each other, is ultimately easier to understand.

I agree that it would be ideal to be able to provide an interface which clearly supports every useful `source` -> `sink` combination in a clear way. I am happy to work with you on the next iteration of this interface.
 - Ideas:
   - Support a plug-in to override aspects of the interface
   - Expose currently unimplemented source->sink combinations
   - Wholesale revamp incl. drag and drop of outfit onto graphical representation of ship (e.g. assign gun to specific port), display of available space, highlight outfits according to if they'll fit or not, etc.

But right now, after 14 months of effort, we finally have something that is better than it was and addresses many issues with a few minor limitations that have work-arounds.

**Comment by ThrawnCA** [comment]
Rather than multiple keys, the same effect could be produced by cleaning up the behavior of the existing checkboxes.

Check "cargo", the buttons become Buy to Cargo and Sell from Cargo, and the product list shows items in the outfitter or cargo. Check "storage", they become Buy to Storage and Sell from Storage, product list shows outfitter+storage. Check "cargo" and "installed", they become Install from Cargo and Uninstall to Cargo, product list shows items in cargo or installed.

The point is to have some way of switching modes and then offer two opposite actions.

> What's your keyboard shortcut for "Buy" then?

'b', same as before.
